src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,105 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
+
 public class ArrayList<T> implements List<T> {
 
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private int currentSize;
+    private T[] values;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        values[currentSize++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAddAndRemove(index);
+        ensureCapacity();
+        System.arraycopy(values, index, values, index + 1, currentSize - index);
+        values[index] = value;
+        currentSize++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGetAndSet(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexForGetAndSet(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForAddAndRemove(index);
+        T valueToReturn = values[index];
+        int toCopy = currentSize - index - 1;
+        System.arraycopy(values, index + 1, values, index, toCopy);
+        values[--currentSize] = null;
```

Comment:
Changed from 
1) `values[currentSize--] = null;`
to
2) `values[--currentSize] = null;`
But I don't understand why.

Reply:
if size is `n`, what will be the value of the last index?

I would suggest you to debug this method in both cases and look what is happening
if it will not be clear after that, please be free to ask again, ok?

Reply:
I checked it with debugger, but still don't understand.
It does the opposite of what I expected.